test/batcher/batcher_test.exs

```diff
@@ -311,6 +312,369 @@ defmodule BorsNG.Worker.BatcherTest do
+  end
+
+  test "rejects a patch with missing require reviewers - using wildcard in CODEOWNERS", %{proj: proj} do
+
+    GitHub.ServerMock.put_state(%{
+      {{:installation, 91}, 14} => %{
+        branches: %{},
+        commits: %{},
+        comments: %{1 => []},
+        teams: %{
+          "my_org" => %{
+            "my_team" => %{}
+          }
+        },
+        statuses: %{"Z" => %{"cn" => :ok}},
+        pulls: %{
+          1 => %Pr{
+            number: 1,
+            title: "Test",
+            body: "Mess",
+            state: :open,
+            base_ref: "master",
+            head_sha: "00000001",
+            head_ref: "update",
+            base_repo_id: 14,
+            head_repo_id: 14,
+            merged: false
+          }
+        },
+        reviews: %{1 => %{"APPROVED" => 0, "CHANGES_REQUESTED" => 0, "approvers" => []}},
+        files: %{"Z" => %{"bors.toml" =>
+          ~s"""
+          status = [ "ci" ]
+          pr_status = [ "cn" ]
+          use_codeowners = true
+          """},
+          "master" => %{".github/CODEOWNERS" =>
+            ~s"""
+            *.toml               @my_org/my_team
+            """},
+        },
+      }})
+    patch = %Patch{
+              project_id: proj.id,
+              pr_xref: 1,
+              commit: "Z",
+              into_branch: "master"}
+            |> Repo.insert!()
+    Batcher.handle_cast({:reviewed, patch.id, "rvrr"}, proj.id)
+    state = GitHub.ServerMock.get_state()
+    assert state == %{
+             {{:installation, 91}, 14} => %{
+               branches: %{},
+               commits: %{},
+               teams: %{
+                 "my_org" => %{
+                   "my_team" => %{}
+                 }
+               },
+               pulls: %{
+                 1 => %Pr{
+                   number: 1,
+                   title: "Test",
+                   body: "Mess",
+                   state: :open,
+                   base_ref: "master",
+                   head_sha: "00000001",
+                   head_ref: "update",
+                   base_repo_id: 14,
+                   head_repo_id: 14,
+                   merged: false
+                 }
+               },
+               comments: %{
+                 1 => [":-1: Rejected because of missing code owner approval"]},
+               statuses: %{"Z" => %{"cn" => :ok}},
+               files: %{"Z" => %{"bors.toml" =>
+                 ~s"""
+                 status = [ "ci" ]
+                 pr_status = [ "cn" ]
+                 use_codeowners = true
+                 """},
+                 "master" => %{".github/CODEOWNERS" =>
+                   ~s"""
+                   *.toml               @my_org/my_team
+                   """},
+               },
+               reviews: %{1 => %{"APPROVED" => 0, "CHANGES_REQUESTED" => 0, "approvers" => []}},
+             }}
+    # When preflight checks reject a patch, no batch should be created!
+    assert [] == Repo.all(Batch)
+  end
```

Comment:
Thank you very much for writing test cases for this part as well. I know what a pain these "big integration tests(tm)" are to work on, since they're running half the code base, but I find that they tend to find bugs.

Reply:
They work but they so hard to read and reason about.  You need a really deep understanding of what GitHub is doing to really get it.